src/AuthRoles.js

```diff
@@ -0,0 +1,322 @@
+const RestQuery = require('./RestQuery');
+import  _ from "lodash";
+import Auth from "./Auth";
+
+// operation result for role
+const OppResult = {
+  rejected: 0,      // role rejected (no path to role was valid)
+  accepted: 1,      // role accepted (at least one path to role was valid)
+  inconclusive: 2,  // circular
+  processing: 3     // role is being validated (this prevents circular roles)
+}
+
+// add only uniquely to array
+const addIfNeed = function(array, string){
+  if(_.indexOf(array, string) === -1){
+    array.push(string)
+  }
+}
+
+export function AuthRoles(auth: Auth, master, isMaster = false) {
+  this.auth = auth
+  this.userId = auth.user ? auth.user.id : undefined
+  this.master = master
+  this.isMaster = isMaster;
+  // manifest contains each role
+  // keyed to its objectId for fast access
+  // { id: { roleInfo } }
+  // roleInfo :
+  // - name: role name
+  // - objectId: role id
+  // - parents[]: the available paths to this role, since
+  //              some roles can be accessed via multiple paths.
+  //              Paths are simply a way to have access to the role by hierarchy aka'<Role Relation>'
+  // - tag (OppResult)
+  this.manifest = {}
+  // array of objectIds pending computation
+  this.toCompute = []
+  // array of objectIds already computed
+  this.computed = []
+  // array of role names and ids the user have access to
+  this.accessibleRoles = { ids: [],  names: [] }
+  // Contains the role that is blocking another role.
+  // It is used to quicky re-accept roles that were previously
+  // rejected because another role has been rejected or inconclusive.
+  // { roleBlocking : [ rolesBlocked ] }
+  this.rejections = {}
+}
+
+// returns a promise that resolves with 'accessibleRoles'
+// once all roles are computed
+AuthRoles.prototype.findRoles = function() {
+  return this.findDirectRoles()
+    .then(() => this.findRolesOfRolesRecursively())
+    .then(() => this.computeAccess())
+    .then(() => this.cleanup())
+    .then(() => Promise.resolve(this.accessibleRoles))
+}
+
+// Note: not sure if this is needed
+AuthRoles.prototype.cleanup = function() {
+  delete this.manifest
+  delete this.toCompute
+  delete this.computed
+  delete this.rejections
+}
+
+/**
+ * Resolves with a promise once all direct roles are fetched.
+ * Direct roles are roles the user is in the 'users' relation.
+ */
+AuthRoles.prototype.findDirectRoles = function() {
+  var restWhere = { 'users': {  __type: 'Pointer', className: '_User', objectId: this.userId } };
+  var query = new RestQuery(this.auth.config, this.master, '_Role', restWhere, {})
+  return query.execute()
+    .then((response) => {
+      var directRoles = response.results
+      this.addToManifest(directRoles)
+    })
+}
+
+/**
+ * Resolves with a promise once all roles inherited by a role are fetched.
+ * Inherited roles are roles a single role has access to by the 'roles' relation.
+ */
+AuthRoles.prototype.findRolesOfRolesRecursively = function() {
```

Comment:
We prefer algorithms without side effects instead of mutations of the global object state: ie: instead of storing the properties in the global state object, store the result of each operation locally, and then return the final result after the recursion for example:
```
function somethingResursive(currentState = {}) {
   if (breakingCondition) {
     return currentState;
   }
   const result = doSomething():
   const mergedResult = mergeStates(currentState, result);
   return somethingResursive(mergedResult);
}
```

Reply:
Alright, let me see what I can do about this.

Reply:
It’s cleaner and more testable usually :) also, please use ES6 classes now :)